posts/2022-11-03-Rust-1.65.0.md

```diff
@@ -0,0 +1,177 @@
+diverging `else` block that executes when that pattern doesn't match.
+
+    let PATTERN: TYPE = EXPRESSION else DIVERGING_BLOCK;
+
+Normal `let` statements can only use _irrefutable_ patterns, statically known
+to always match. That pattern is often just a single variable binding, but may
+also unpack compound types like structs, tuples, and arrays. However, that was
+not usable for conditional matches, like pulling out a variant of an enum --
+until now! With `let`-`else`, a refutable pattern can match and bind variables
+in the surrounding scope like a normal `let`, or else diverge (e.g. `break`,
+`return`, `panic!`) when the pattern doesn't match.
+
+```rust
+fn get_count_item(s: &str) -> (u64, &str) {
+    let mut it = s.split(' ');
+    let (Some(count_str), Some(item)) = (it.next(), it.next()) else {
+        panic!("Can't segment count item pair: '{s}'");
+    };
+    let Ok(count) = u64::from_str(count_str) else {
+        panic!("Can't parse integer: '{count_str}'");
+    };
+    (count, item)
+}
+assert_eq!(get_count_item("3 chairs"), (3, "chairs"));
+```
+
+The scope of name bindings is the main thing that makes this different from
+`match` or `if let`-`else` expressions. You could previously approximate these
+patterns with an unfortunate bit of repeation to an outer `let`:
+
+```rust
+    let (count_str, item) = match (it.next(), it.next()) {
+        (Some(count_str), Some(item)) => (count_str, item),
+        _ => panic!("Can't segment count item pair: '{s}'"),
+    };
+    let count = if let Ok(count) = u64::from_str(count_str) {
+        count
+    } else {
+        panic!("Can't parse integer: '{count_str}'");
+    };
+```
+
+### `break` from labeled blocks
+
+Plain block expressions can now be labeled as a `break` target, terminating
+that block early. This may sound a little like a `goto` statement, but it's not
+an arbitrary jump, only from within a block to its end. This was already
+possible with `loop` blocks, and you may have seen people write loops that
+always execute just once, only to get a labeled `break`.
+
+Now there is a direct language feature for that! Labeled `break` may also
+include an expression value, just as with loops, letting a multi-statement
+block have an early "return" value.
+
+```rust
+let result = 'block: {
+    do_thing();
+    if condition_not_met() {
+        break 'block 1;
+    }
+    do_next_thing();
+    if condition_not_met() {
+        break 'block 2;
+    }
+    do_last_thing();
+    3
+};
+```
+
+### Splitting Linux debuginfo
+
+Back in Rust 1.51, the compiler team added support for [split debug
+information](https://blog.rust-lang.org/2021/03/25/Rust-1.51.0.html#splitting-debug-information)
+on macOS with `-Csplit-debuginfo=unpacked`, and now this option is stable for
+use on Linux as well, producing separate DWARF `.dwo` object files. This
+separation may improve build times, especially for the linker as it will simply
+have less data to process.
```

Comment:
This sentence is pretty confusing to me, and possibly to other readers as well. Why does it have less data to process now, isn't it just split and therefore the same, but in different files? I'm not sure how that makes anything be processed faster.

Reply:
It's possible that I'm misunderstanding or misrepresenting it too -- @davidtwco can you help clarify expectations here?

Reply:
Something like this, maybe?

> Back in Rust 1.51, the compiler team stabilized support for [split debug information](https://blog.rust-lang.org/2021/03/25/Rust-1.51.0.html#splitting-debug-information) on macOS, and now this option is stable for use on Linux as well. 
>
> `-Csplit-debuginfo=unpacked` will split debuginfo out into multiple `.dwo` DWARF object files and `-Csplit-debuginfo=packed` will produce a single `.dwp` DWARF package alongside your output binary with all the debuginfo packaged together.
>
> Split DWARF lets the linker avoid processing the debuginfo (because it isn't in the object files being linked anymore), which can speed up link times!
```

Reply:
Thanks!